### ISSUE TITLE
glslify-sync

### ISSUE THREAD

**Issue by vorg** [stackgl/glslify#61]
After discussing glslify usage in node here https://github.com/stackgl/glslify/issues/8 and here https://github.com/stackgl/glslify/issues/47  I've been using [glslify-promise](https://github.com/stackgl/glslify-promise) for a while now. While it solves problem of clean a API in nodejs there are problems once you start using it in 3rd party modules. 

I.e. my main app can wait for resources to load but if I require e.g. `var downsampleCubemap = require('cubemap-downsampler')` I can't wait until some internal shader promise resolves. Of course I could make `downsampleCubemap` return a promise too but then we have promise pollution and it doesn't solve problem that I need to exectue GPU command sequentially within one frame.

Therefore **glslify-sync** On the API level it works exactly the same as `glslify-promise` but internally I'm using `child_process/execSync`. Yes, it's blocking but i find it cleaner that promises everywhere or using glslify to preprocess glsl files and use `brfs` to inject bunch of *.out.frag and *.out.vert files.

In JS:
```javascript
var glslifySync = require('glslify-sync');
var VERT = glslifySync(__dirname + '/glsl/rect.vert');
````

For Browserify I copy/pasted glslify-sync/transform, so:
```
browserify main.js -g glslify-sync/transform
```

I'm posting it here before I push to npm to see if there is any conflict of interest (ideology) and if we could actually host it here at stackgl or should i push and maintain it myself.

**Comment by hughsk** [comment]
Very happy to see this, and happy to host it on stackgl provided nobody else objects :) Would you like me to add you to the org so we can get that set up?

**Comment by vorg** [comment]
@hughsk that would be great!

**Comment by hughsk** [comment]
Sweet, done! :grin:

**Comment by mattdesl** [comment]
:tada: Cool.

@hughsk Why is child_process and/or async needed? I'm wondering for runtime purposes: #58 #12.

**Comment by hughsk** [comment]
@mattdesl accessing the file system is done asynchronously for performance when glslify is run on a server or Electron app :)

[glslbin's source](https://github.com/stackgl/glslbin) has some tricks in there for runtime shader generation backed by a server, but it could be repurposed to do it all on the client with the right build process!

**Comment by vorg** [comment]
I haven't pushed it yet as I have issues with this approach. In my poject 14x `execSync` is 3 times slower than using `glslify-promise` probably due to duplicated work of traversing the folder while looking for dependencies.

**Comment by vorg** [comment]
For the current project i cache the results of glslify-ing `assets/glsl/lightShader.vert` in  `assets/glsl/.lightShader.vert` and rebuild if the file was modified. So now my load time goes to `0s` on the second run :). Of course that's very naive approach and I should check the time of the dependencies too...

**Comment by vorg** [comment]
I've implemented a very naive version with `execSync` https://www.npmjs.com/package/glslify-sync/
It works but gets slow with multiple shaders as you run glslify multiple times. I did it as I need to distribute shaders within a library and I expect them to be available immediately in node. Another solution would be to precompile them and distribute full glslify-ied versions together with `brfs` transform for the web.
